### PR TITLE
test: migrated custom-parser.3.test.js from tap to node:test

### PR DESCRIPTION
**Pull request by Tony133**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

Proposal:

 - migrated `custom-parser.3.test.js` from `tap` to `node:test` 🔥

### CODE REVIEW

**Review by gurgunday — COMMENTED** [review]
Please address @dancastillo's feedback

File: test/custom-parser.3.test.js [hunk 1]
```diff
@@ -74,14 +71,15 @@ test('contentTypeParser should add a custom parser with RegExp value', t => {
...
-        t.equal(response.statusCode, 200)
-        t.same(body.toString(), JSON.stringify({ hello: 'world' }))
+        t.assert.ifError(err)
+        t.assert.strictEqual(response.statusCode, 200)
+        t.assert.deepStrictEqual(body.toString(), JSON.stringify({ hello: 'world' }))
       })
     })
 
-    t.test('in OPTIONS', t => {
+    t.test('in OPTIONS', (t) => {
```
[the remark this answers is not in the document]
Comment (Tony133) [reply test/custom-parser.3.test.js]: Done see here: [ab818bb](https://github.com/fastify/fastify/pull/5903/commits/ab818bbbefaec9db83a7465c43ea3c1fb20f83a3)

File: test/custom-parser.3.test.js [hunk 2]
```diff
@@ -59,12 +57,11 @@ test('contentTypeParser should add a custom parser with RegExp value', t => {
   })
 
   fastify.listen({ port: 0 }, err => {
-    t.error(err)
-
-    t.teardown(() => fastify.close())
+    t.assert.ifError(err)
 
-    t.test('in POST', t => {
+    t.test('in POST', (t) => {
```
[the remark this answers is not in the document]
Comment (Tony133) [reply test/custom-parser.3.test.js]: Done see here: [ab818bb](https://github.com/fastify/fastify/pull/5903/commits/ab818bbbefaec9db83a7465c43ea3c1fb20f83a3)

File: test/custom-parser.3.test.js [hunk 3]
```diff
@@ -31,18 +30,17 @@ test('should be able to use default parser for extra content type', t => {
...
+      t.assert.deepStrictEqual(JSON.parse(body.toString()), { hello: 'world' })
+      done()
     })
   })
 })
 
-test('contentTypeParser should add a custom parser with RegExp value', t => {
-  t.plan(3)
-
+test('contentTypeParser should add a custom parser with RegExp value', (t) => {
```
[the remark this answers is not in the document]
Comment (Tony133) [reply test/custom-parser.3.test.js]: Done see here: [ab818bb](https://github.com/fastify/fastify/pull/5903/commits/ab818bbbefaec9db83a7465c43ea3c1fb20f83a3)

File: test/custom-parser.3.test.js [hunk 4]
```diff
@@ -58,13 +56,12 @@ test('contentTypeParser should add a custom parser with RegExp value', t => {
...
 
-  fastify.listen({ port: 0 }, err => {
-    t.error(err)
-
-    t.teardown(() => fastify.close())
+  fastify.listen({ port: 0 }, async err => {
+    t.assert.ifError(err)
 
-    t.test('in POST', t => {
+    await t.test('in POST', (t) => {
```
Comment (dancastillo) [inline test/custom-parser.3.test.js:62]: we would still need to add in `done` and call after last assert to ensure the planned asserts were run
```suggestion
    await t.test('in POST', (t, done) => {
```

File: test/custom-parser.3.test.js [hunk 5]
```diff
@@ -74,14 +71,15 @@ test('contentTypeParser should add a custom parser with RegExp value', t => {
...
-        t.equal(response.statusCode, 200)
-        t.same(body.toString(), JSON.stringify({ hello: 'world' }))
+        t.assert.ifError(err)
+        t.assert.strictEqual(response.statusCode, 200)
+        t.assert.deepStrictEqual(body.toString(), JSON.stringify({ hello: 'world' }))
       })
     })
 
-    t.test('in OPTIONS', t => {
+    await t.test('in OPTIONS', (t) => {
```
Comment (dancastillo) [inline test/custom-parser.3.test.js:80]: we would still need to add in `done` and call after last assert
```suggestion
    await t.test('in POST', (t, done) => {
```

Comment (Tony133) [reply test/custom-parser.3.test.js]: Done! see here: [8ccbf71](https://github.com/fastify/fastify/pull/5903/commits/8ccbf71c4bfddae19cd2e07926a58ade1f610813)

File: test/custom-parser.3.test.js (hunk 4 above)
Comment (Tony133) [reply test/custom-parser.3.test.js]: Done! see here: [8ccbf71](https://github.com/fastify/fastify/pull/5903/commits/8ccbf71c4bfddae19cd2e07926a58ade1f610813)

File: test/custom-parser.3.test.js [hunk 6]
```diff
@@ -237,9 +237,10 @@ test('catch all content type parser should not interfere with content type parse
         'Content-Type': 'text/html'
       }
     }, (err, response, body) => {
-      t.error(err)
-      t.equal(response.statusCode, 200)
-      t.same(body.toString(), 'my texthtml')
+      t.assert.ifError(err)
+      t.assert.strictEqual(response.statusCode, 200)
+      t.assert.deepStrictEqual(body.toString(), 'my texthtml')
+      done()
```
Comment (Eomm) [inline test/custom-parser.3.test.js:243]: When we run multiple calls in parallel (no await), we must wait for them all to be completed before calling `done`.

In this scenario, the last sget may end before the other 2, and the test would complete anyway.

Example solution: https://github.com/fastify/fastify/pull/5881/files#diff-6a98dbdd7ca9fb8e0adfcc8787c950f65344cd9581437118754f4adb1a791f7d

Comment (Tony133) [reply test/custom-parser.3.test.js:243]: If I understand correctly, I have to add this:
```ts
      let pending = 2
      function completed () {
        if (--pending == 0) {
          done()
        }
      }
```
correct ?

Comment (gurgunday) [reply test/custom-parser.3.test.js:243]: In parallel, that would work

**Review by gurgunday — APPROVED** [review]
lgtm

**Review by Eomm — APPROVED** [review]

**Pull request merged**